### ISSUE TITLE
Rule S3963: to accept finalizer case

### ISSUE THREAD

**Issue by am11** [SonarSource/sonar-dotnet#2247]
### Description

In my understanding, if the `static` field is initialized with property of a disposable object within a finalizer block in cctor, it should not be flagged as a violation of rule S3963.

### Repro steps

```c#
class Program
{
  private static readonly string sm_processName;
  // produces S3963 violation
  static Program()
  {
    using (var process = Process.GetCurrentProcess())
    {
      sm_processName = process.ProcessName;
    }
  }

  static void Main()
  {
    Console.WriteLine(sm_processName);
  }
}
```

### Expected behavior

Do not violate S3963.

### Actual behavior

Violates S3963.

### Known workarounds

Use a static method:

```diff
- private static readonly string sm_processName;
- static Program()
+ private static readonly string sm_processName = GetProcessName();
+ static string GetProcessName()
  {
    using (var process = Process.GetCurrentProcess())
    {
-      sm_processName = process.ProcessName;
+      return process.ProcessName;
    }
  }
```

### Related information

* SonarLint 4.8.0.4040
* Visual Studio 2017

**Comment by Evangelink** [comment]
I am closing this thread because I don't think there is any False Positive here.

**Comment by am11** [comment]
Thank you for the reply and sharing details @valhristov (and sorry for not getting back earlier).
TIL: going by CA1810, the magic really happens with `beforefieldinit`, otherwise IL-wise, an implicit `.cctor` added in case of no explicit static constructor actually produces more code.

<details>
  <summary><ins>Differential of <a href="https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACh8jiA6AETQgHMA7Ae1jQGMoLqAJloB2KgG8qBAgAcEaAG4R4tAGwEEcCABMObDAE9ahKGAD6cjjzhQoAOQhg4A6QHpXshB20BXa1AIAZQBmAE5VYIIFNA4MZRi2KTVRAAoASiTJSmlpGhSlBE8rGwCAXgIABS9/KDoAcTgYAGEfBE02GCri23SM7JysnJzTC2qShycCcssaui6aieck6QBfJLXKJJJ1HAAWAgBZCDQ2dMzl2lCUkZnxxzg0lwINlaA"><code>with explicit cctor</code></a> vs. <a href="https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACh8jiA6AETQgHMA7Ae1jQGMoLqAJloB2KgG8qBAgAcEaAG4R4tAGwEEcCABMObDAE9ahKGAD6cjjzhQoAOQhg4BALwEAsgZIiAFAEoBaQB6INkEDm0AV2soAgBlAGYATlUEggU0DgxlTLYpNWMPL2JfP3zJSmlpGh8lBDCrG1i3AAVwmKg6AHE4GABhSIRNNhg2xtt/MsqqiqqqnBEGjroxjocnQKqAX3ydynySdRwAFg8INDZ/cvzq4iSfUwt2pvW4AN2qLaA=) (click or tap to expand"><code>without explicit cctor</code></a></ins></summary>
<br />

```diff
 .class private auto ansi '<Module>'
 {
 } // end of class <Module>
 
-.class private auto ansi Program
+.class private auto ansi beforefieldinit Program
     extends [mscorlib]System.Object
 {
     // Fields
     .field private static initonly string sm_processName
 
     // Methods
-    .method private hidebysig specialname rtspecialname static 
-        void .cctor () cil managed 
+    .method private hidebysig static 
+        string MyProgram () cil managed 
     {
         // Method begins at RVA 0x2050
-        // Code size 30 (0x1e)
+        // Code size 27 (0x1b)
         .maxstack 1
         .locals init (
-            [0] class [System]System.Diagnostics.Process
+            [0] class [System]System.Diagnostics.Process,
+            [1] string
         )
 
         IL_0000: call class [System]System.Diagnostics.Process [System]System.Diagnostics.Process::GetCurrentProcess()
         IL_0005: stloc.0
         .try
         {
             IL_0006: ldloc.0
             IL_0007: callvirt instance string [System]System.Diagnostics.Process::get_ProcessName()
-            IL_000c: stsfld string Program::sm_processName
-            IL_0011: leave.s IL_001d
+            IL_000c: stloc.1
+            IL_000d: leave.s IL_0019
         } // end .try
         finally
         {
             // sequence point: hidden
-            IL_0013: ldloc.0
-            IL_0014: brfalse.s IL_001c
+            IL_000f: ldloc.0
+            IL_0010: brfalse.s IL_0018
 
-            IL_0016: ldloc.0
-            IL_0017: callvirt instance void [mscorlib]System.IDisposable::Dispose()
+            IL_0012: ldloc.0
+            IL_0013: callvirt instance void [mscorlib]System.IDisposable::Dispose()
 
             // sequence point: hidden
-            IL_001c: endfinally
+            IL_0018: endfinally
         } // end handler
 
-        IL_001d: ret
-    } // end of method Program::.cctor
+        IL_0019: ldloc.1
+        IL_001a: ret
+    } // end of method Program::MyProgram
 
     .method private hidebysig static 
         void Main () cil managed 
     {
-        // Method begins at RVA 0x208c
+        // Method begins at RVA 0x2088
         // Code size 11 (0xb)
         .maxstack 8
 
         IL_0000: ldsfld string Program::sm_processName
         IL_0005: call void [mscorlib]System.Console::WriteLine(string)
         IL_000a: ret
     } // end of method Program::Main
 
     .method public hidebysig specialname rtspecialname 
         instance void .ctor () cil managed 
     {
-        // Method begins at RVA 0x2098
+        // Method begins at RVA 0x2094
         // Code size 7 (0x7)
         .maxstack 8
 
         IL_0000: ldarg.0
         IL_0001: call instance void [mscorlib]System.Object::.ctor()
         IL_0006: ret
     } // end of method Program::.ctor
 
+    .method private hidebysig specialname rtspecialname static 
+        void .cctor () cil managed 
+    {
+        // Method begins at RVA 0x209c
+        // Code size 11 (0xb)
+        .maxstack 8
+
+        IL_0000: call string Program::MyProgram()
+        IL_0005: stsfld string Program::sm_processName
+        IL_000a: ret
+    } // end of method Program::.cctor
+
 } // end of class Program
```
</details>
<br />

as per the docs, in both cases we get the `void .cctor () cil managed` but the key difference is presence of `beforefieldinit`.
